api/http-api.md

```diff
@@ -0,0 +1,176 @@
+# HTTP API
+
+## Overview
+
+Temporal has a gRPC API but no HTTP API. We need an HTTP API.
+
+**Goals**
+
+* Create HTTP API that governs all reasonable user-facing calls
+* Ensure new API calls are automatically adopted
+* Expose HTTP API in server by default for all to use like gRPC, but have option to disable
+* Support all existing interceptors and auth methods
+
+**Non-Goals**
+
+* Design a high-quality REST API from first principles
+* grpc-web (browser-only gRPC workaround that requires server-side support)
+* Make an easy call to get workflow result
+
+## Approach
+
+This project will use https://github.com/grpc-ecosystem/grpc-gateway and annotations inspired by work already done on
+the UI server for similar purpose at https://github.com/temporalio/ui-server. After review of the UI server's
+annotations, they so closely match our goals that many things won't change.
+
+Since namespace is required for most calls and is usually URL safe, we have decided to put the namespace in the URL
+instead of as an HTTP header. This makes URL's large and a bit unwieldy, but calls like from `curl` would have to have
+`-H Temporal-Namespace: <my-ns>` which is even more unwieldy.
+
+The bodies of requests are proto JSON of the request messages and responses are full proto JSON of the response
+messages. While we try to somewhat follow REST principles, this is not a full RESTful API necessarily. Developers adding
+new API will need to add the gRPC gateway annotations for their RPC calls.
+
+For the initial version, we will only be doing workflow service. Operator service and cloud APIs may come soon after
+though.
+
+### Rules
+
+* Naming
+  * Multi-word URL paths use kebab-case
+  * Multi-word query parameters use camelCase (only first letter of abbreviations capitalized if not first word)
+  * Do not assume any names reserved. So we can't have `/workflows/{workflow_id}` and `/workflows/general-call` in case
+    there's a workflow named `general-call`.
+  * Use plural nouns where an identifier may be provided to differentiate
+    * But singular nouns are not _required_ in other places, can still be plural
+* Response codes
+  * 201 success if a creation
+  * 202 if action eventually consistent
+  * 200 other all other successes
+  * Do not use 204 for no response, even if there's no response because we may add response later, just 200 everything
+* Methods:
+  * `GET` if read-only
+    * Noun as last URL term
+    * Use query parameters for commonly set items to avoid relying on request bodies (but still rely on bodies as
+      needed). This may require some gRPC gateway customization.
```

Comment:
It will also be hard to generate type safe clients if we break the request params out of proto messages.

Reply:
Looks like we're in for a lot of hand coding.

Reply:
I'm wondering how complex are read only request types are, might not be a big issue.

Reply:
> It will also be hard to generate type safe clients if we break the request params out of proto messages.

This is already happening with URL path members coming from proto messages.

> Looks like we're in for a lot of hand coding.

gRPC gateway mostly supports this out of the box. Though they default to letting _every_ field of a `GET` message be in the query params if no body is specified. There may be some hand-coding here, but we only have to do it once and how I would do it is via Temporal-based proto annotations. Not per RPC code.

> I'm wondering how complex are read only request types are, might not be a big issue.

Not too bad, but I have to investigate whether we can choose specific query params or if we have to choose "query or body" and that's all. Either way, anything I do here will be generally applicable, so worst case you'd have to specify which body members can be part of the query.

Reply:
> gRPC gateway mostly supports this out of the box
We'll have to make sure we keep the TS client in sync with each of these deviations from the proto definitions.

Reply:
Yes, the HTTP API is not 100% machine derived from the gRPC API. People (internal or users) that want to use the APIs will need to know what to call and with what options. Also notice the URL names don't match the gRPC names. You'll have to keep that in sync too. Basically we're going to API maintenance simplicity (i.e. not a bunch of custom HTTP-server-layer code) while still doing the best we can with HTTP constructs.

Reply:
I think it's fine, it should be fairly straightforward to translate between HTTP and gRPC.

Reply:
Resolving